handlers.go

```diff
@@ -1,24 +1,169 @@
 package nitriding
 
 import (
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
 	"fmt"
+	"io"
 	"net/http"
+	"net/url"
+	"strings"
 )
 
-func formatIndexPage(appURL string) string {
-	page := "This host runs inside an AWS Nitro Enclave.\n"
-	if appURL != "" {
+const (
+	// The maximum length of the key material (in bytes) that enclave
+	// applications can PUT to our HTTP API.
+	maxKeyMaterialLen = 1024 * 1024
+	// The HTML for the enclave's index page.
+	indexPage = "This host runs inside an AWS Nitro Enclave.\n"
+)
+
+var (
+	errFailedReqBody  = errors.New("failed to read request body")
+	errFailedGetState = errors.New("failed to retrieve saved state")
+	errNoAddr         = errors.New("parameter 'addr' not found")
+	errHashWrongSize  = errors.New("given hash is of invalid size")
+)
+
+func formatIndexPage(appURL *url.URL) string {
+	page := indexPage
+	if appURL != nil {
 		page += fmt.Sprintf("\nIt runs the following code: %s\n"+
 			"Use the following tool to verify the enclave: "+
-			"https://github.com/brave-experiments/verify-enclave", appURL)
+			"https://github.com/brave-experiments/verify-enclave", appURL.String())
 	}
 	return page
 }
 
-// getIndexHandler returns an index handler that informs the visitor that this
-// host runs inside an enclave.
-func getIndexHandler(cfg *Config) http.HandlerFunc {
+// rootHandler returns a handler that informs the visitor that this host runs
+// inside an enclave.  This is useful for testing.
+func rootHandler(cfg *Config) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
 		fmt.Fprintln(w, formatIndexPage(cfg.AppURL))
 	}
 }
+
+// reqSyncHandler returns a handler that lets the enclave application request
+// state synchronization, which copies the given remote enclave's state into
+// our state.
+//
+// FIXME: https://github.com/brave/nitriding/issues/44
+func reqSyncHandler(e *Enclave) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		q := r.URL.Query()
+		// The 'addr' parameter must have the following form:
+		// https://example.com:8443
+		addrs, ok := q["addr"]
+		if !ok {
+			http.Error(w, errNoAddr.Error(), http.StatusBadRequest)
+			return
+		}
+		addr := addrs[0]
+
+		// Are we dealing with a well-formed URL?
+		if _, err := url.Parse(addr); err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		if err := RequestKeys(addr, e.KeyMaterial); err != nil {
+			http.Error(w, fmt.Sprintf("failed to synchronize state: %v", err), http.StatusInternalServerError)
+			return
+		}
+		w.WriteHeader(http.StatusOK)
+	}
+}
+
+// getStateHandler returns a handler that lets the enclave application retrieve
+// previously-set state.
+func getStateHandler(e *Enclave) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "application/octet-stream")
+		s, err := e.KeyMaterial()
```

Comment:
Key material can be set in two ways:
1. Within the enclave, by the trusted enclave application. In this case, we don't need validation because all involved parties are trusted.
2. Across the Internet (or ideally: across our private Kubernetes network), when enclaves synchronize with each other. In this case, we do mutual authentication using attestation documents.

Does that answer your question? I hope I didn't miss anything.